engine/src/main/java/org/hibernate/validator/internal/engine/valueextraction/ValueExtractorManager.java

```diff
@@ -347,6 +351,124 @@ private static boolean isClassPresent(String className, boolean fallbackOnTCCL)
+				valueExtractorDescriptors = getMaximallySpecificValueExtractors(
+						valueType,
+						valueExtractorCandidates
+				);
+				if ( valueExtractorDescriptors.isEmpty() ) {
+					nonContainerTypes.add( cacheKey );
+				}
+				else {
+					valueExtractorCache.put( cacheKey, valueExtractorDescriptors );
+				}
+			}
+
+			return valueExtractorDescriptors;
+		}
+
+//		public Set<ValueExtractorDescriptor> getValueExtractors(Class<?> runtimeType, TypeVariable<?> typeParameter) {
+//			ValueExtractorCacheKey cacheKey = new ValueExtractorCacheKey( runtimeType, typeParameter );
+//			Set<ValueExtractorDescriptor> valueExtractorDescriptors = valueExtractorCache.get( cacheKey );
+//			if ( valueExtractorDescriptors == null ) {
+//				valueExtractorDescriptors = getValueExtractors( cacheKey );
+//				if ( !valueExtractorDescriptors.isEmpty() ) {
+//					valueExtractorCache.put( cacheKey, valueExtractorDescriptors );
+//				}
+//			}
+//
+//			return valueExtractorDescriptors;
+//		}
+//
+//		private Set<ValueExtractorDescriptor> getValueExtractors(ValueExtractorCacheKey key) {
+//			if ( nonContainerTypes.contains( key ) ) {
+//				return Collections.emptySet();
+//			}
+//
+//			boolean isInternal = TypeVariables.isInternal( key.typeParameter );
+//			Map<Class<?>, Map<TypeVariable<?>, TypeVariable<?>>> allBindings = null;
+//			if ( !isInternal ) {
+//				allBindings = TypeVariableBindings.getTypeVariableBindings( (Class<?>) key.typeParameter.getGenericDeclaration() );
+//			}
+//
+//			Set<ValueExtractorDescriptor> typeCompatibleExtractors = valueExtractors.values()
+//					.stream()
+//					.filter( e -> TypeHelper.isAssignable( key.runtimeType, e.getContainerType() ) )
+//					.collect( Collectors.toSet() );
+//
+//			Set<ValueExtractorDescriptor> containerElementCompliantExtractors = new HashSet<>();
+//
+//			for ( ValueExtractorDescriptor extractorDescriptor : typeCompatibleExtractors ) {
+//				TypeVariable<?> typeParameterBoundToExtractorType;
+//
+//				if ( !isInternal ) {
+//					Map<TypeVariable<?>, TypeVariable<?>> bindingsForExtractorType = allBindings.get( extractorDescriptor.getContainerType() );
+//					typeParameterBoundToExtractorType = bind( key.typeParameter, bindingsForExtractorType );
+//				}
+//				else {
+//					typeParameterBoundToExtractorType = key.typeParameter;
+//				}
+//
+//				if ( Objects.equals( extractorDescriptor.getExtractedTypeParameter(), typeParameterBoundToExtractorType ) ) {
+//					containerElementCompliantExtractors.add( extractorDescriptor );
+//				}
+//			}
+//
+//			if ( containerElementCompliantExtractors.isEmpty() ) {
+//				nonContainerTypes.add( key );
+//			}
+//			return getMaximallySpecificValueExtractors( key.runtimeType, containerElementCompliantExtractors );
+//		}
+
+		private class ValueExtractorCacheKey {
+
+			private final Class<?> runtimeType;
+			private final TypeVariable<?> typeParameter;
+
+			ValueExtractorCacheKey(Class<?> runtimeType, TypeVariable<?> typeParameter) {
```

Comment:
Wondering if just having the `TypeVariable<?>`as a key wouldn't be sufficient as it contains the Class too. In my experience `TypeVariable` hashCode and equals are quite slow so maybe we still need to encapsulate it in an object to cache the hashCode (one of my project is to get rid of `TypeVariable` and have a `TypeParameterDescriptor` similar to our `AnnotationDescriptor` but let's keep that for later).